src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -0,0 +1,420 @@
+    Transaction& t,
+    std::list<LogicalCachedExtentRef>& extents) final {
+    return seastar::do_with(
+      std::map<Writer*, std::list<LogicalCachedExtentRef>>(),
+      [this, extents=std::move(extents), &t](auto& alloc_map) {
+      for (auto& extent : extents) {
+        auto writer = &(get_writer(extent->hint));
+        alloc_map[writer].emplace_back(extent);
+      }
+      return trans_intr::do_for_each(alloc_map, [this, &t](auto& p) {
+        auto writer = p.first;
+        auto& extents_to_pesist = p.second;
+        return writer->write(t, extents_to_pesist);
+      });
+    });
+  }
+
+  stop_ertr::future<> stop() {
+    return crimson::do_for_each(writers, [](auto& writer) {
+      return writer.stop();
+    });
+  }
+private:
+  SegmentProvider& segment_provider;
+  SegmentManager& segment_manager;
+  std::vector<Writer> writers;
+  LBAManager& lba_manager;
+  Journal& journal;
+};
+
+class ExtentPlacementManager {
+public:
+  ExtentPlacementManager(
+    Cache& cache,
+    LBAManager& lba_manager
+  ) : cache(cache), lba_manager(lba_manager) {}
+
+  /**
+   * alloc_new_extent_by_type
+   *
+   * Create a new extent, CachedExtent::poffset may not be set
+   * if a delayed allocation is needed.
+   */
+  CachedExtentRef alloc_new_extent_by_type(
+    Transaction& t,
+    extent_types_t type,
+    segment_off_t length,
+    ool_placement_hint_t hint = ool_placement_hint_t::NONE) {
+    // only logical extents should fall in this path
+    assert(is_logical_type(type));
+    auto dtype = get_allocator_type(hint);
+    CachedExtentRef extent;
+    // for extents that would be stored in NVDIMM/PMEM, no delayed
+    // allocation is needed
+    if (need_delayed_allocation(dtype)) {
+      // set a unique temperary paddr, this is necessary because
+      // transaction's write_set is indexed by paddr
+      extent = cache.alloc_new_extent_by_type(
+        t, type, length, delayed_temp_paddr(fake_segment_off));
+      fake_segment_off += length;
+    } else {
+      extent = cache.alloc_new_extent_by_type(t, type, length);
+    }
+    extent->backend_type = dtype;
+    extent->hint = hint;
+    return extent;
+  }
+
+  template<
+    typename T,
+    std::enable_if_t<is_base_of_v<LogicalCachedExtent, T>, int> = 0>
+  TCachedExtentRef<T> alloc_new_extent(
+    Transaction& t,
+    segment_off_t length,
+    ool_placement_hint_t hint = ool_placement_hint_t::NONE)
+  {
+    auto dtype = get_allocator_type(hint);
+    TCachedExtentRef<T> extent;
+    if (need_delayed_allocation(dtype)) {
+      // set a unique temperary paddr, this is necessary because
+      // transaction's write_set is indexed by paddr
+      extent = cache.alloc_new_extent<T>(
+        t, length, delayed_temp_paddr(fake_segment_off));
+      fake_segment_off += length;
+    } else {
+      extent = cache.alloc_new_extent<T>(t, length);
+    }
+    extent->backend_type = dtype;
+    extent->hint = hint;
+    return extent;
+  }
+
```

Comment:
```
/**
 * delayed_transaction_paddr_alloc
 *
 * Performs any outstanding ool writes and updates pending lba updates
 * accordingly
 */

Reply:
Can we rename this to perform_ool_writes?

Reply:
Um...this method also allocate relative addresses for inline extents

Reply:
How about perform_or_inline_ool_writes.  I'd really like the method name to reflect the fact that this call performs writes.

Reply:
ok:-)